File: 141-linked-list-cycle/141-linked-list-cycle.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
    unordered_map<ListNode*, ListNode*> hash;
public:
    bool hasCycle(ListNode *head) {
        ListNode *lst = head;
        while(lst){
            if(hash.find(lst) == hash.end()){
                hash[lst] = lst->next;
            }
            else{
                return true;
            }
            lst = lst->next;
        }
        return false;
    }
};

File: 141-linked-list-cycle/141-linked-list-cycle.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    def hasCycle(self, head: Optional[ListNode]) -> bool:
        if not head:
            return head
        pos = 0
        hashmap = {}
        flag = False
        curr = head
        while curr:
            if curr in hashmap.keys():
                flag = True
                break
            else:
                hashmap[curr] = pos
                pos += 1
            curr = curr.next
        if flag:
            return True
        else:
            return False